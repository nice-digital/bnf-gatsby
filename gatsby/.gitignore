node_modules/
.cache/
public

# dotenv environment variables file
.env
.env.*
