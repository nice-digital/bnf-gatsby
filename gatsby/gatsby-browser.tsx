import {
	ShouldUpdateScrollArgs,
	type RouteUpdateArgs,
	type WrapPageElementBrowserArgs,
} from "gatsby";
import { type ReactElement } from "react";

import { Layout } from "@/components/Layout/Layout";

// Gatsby hook for when the route has changed on the client side
// See https://www.gatsbyjs.org/docs/browser-apis/#onRouteUpdate
export const onRouteUpdate = ({
	prevLocation,
	location,
}: RouteUpdateArgs): void => {
	if (prevLocation) {
		// Cast is needed because of https://github.com/gatsbyjs/gatsby/issues/29124
		const prevPath = (prevLocation as Location).pathname,
			path = location.pathname;

		// Push our own event to the dataLayer on page change rather than using the
		// 'gatsby-route-change' event built into gatsby-plugin-google-tagmanager.
		// Because gatsby-route-change is pushed on initial page load as well as route change,
		// AND because it doesn't use requestAnimationFrame to delay until the page title has been updated

		const sendPageView = () => {
			window.dataLayer.push({
				location: location.href,
				referrer: prevLocation.href,
			});
			// Don't consider hash changes to be a page view - pageviews only happy when the path changes
			if (prevPath != path) window.dataLayer.push({ event: "pageview" });
		};

		// Delay before push to the data layer, to make sure the page title has been updated
		// See https://github.com/gatsbyjs/gatsby/pull/10917/files#diff-bf0d94c8bf47d5c1687e342c2dba1e00R12-R13
		if ("requestAnimationFrame" in window) {
			window.requestAnimationFrame(() => {
				window.requestAnimationFrame(sendPageView);
			});
		} else {
			// simulate 2 rAF calls
			setTimeout(sendPageView, 32);
		}
	} else {
		window.dataLayer.push({
			location: location.href,
			referrer: document.referrer,
		});
	}
	// Default return statement when prevLocation is falsy
	return;
};

export const wrapPageElement = ({
	element,
	props,
}: WrapPageElementBrowserArgs): ReactElement => (
	<Layout {...props}>{element}</Layout>
);

/**
 * Gatsby hook for overriding scroll position
 * See https://www.gatsbyjs.org/docs/browser-apis/#shouldUpdateScroll
 */
export const shouldUpdateScroll = ({
	routerProps: { location },
	prevRouterProps,
	getSavedScrollPosition,
}: ShouldUpdateScrollArgs): boolean | string | [number, number] => {
	if (
		// If there's no previous route props we're coming from an external site, which means
		// we want to scroll to the hash (if there is one), and _not_ a saved scroll position
		!prevRouterProps ||
		// Or we're linking to as hash within the same page
		(prevRouterProps.location.pathname === location.pathname && location.hash)
	) {
		// Provide our own scroll to hash to avoid Gatsby using a stored scroll position
		requestAnimationFrame(() => {
			requestAnimationFrame(() => {
				const targetElement = location.hash
					? document.querySelector(location.hash)
					: null;

				if (targetElement) {
					targetElement.setAttribute("tabIndex", "-1");
					(targetElement as HTMLElement).focus();
					targetElement.scrollIntoView();
				}
			});
		});
		return false;
	}

	const savedScrollY = (getSavedScrollPosition(location)?.[1] || 0) as number;
	if (savedScrollY > 0) {
		window.scrollTo(0, savedScrollY);
		return false;
	}

	if (location.hash) {
		requestAnimationFrame(() => {
			requestAnimationFrame(() => {
				const targetElement = document.querySelector(location.hash);

				// Default to Gatsby's default behaviour if the element doesn't exist
				if (!targetElement) return true;

				// Provide our own scroll to hash to avoid Gatsby using a stored scroll position
				targetElement.setAttribute("tabIndex", "-1");
				(targetElement as HTMLElement).focus();
				targetElement.scrollIntoView();
				return false;
			});
		});
		return false;
	}

	requestAnimationFrame(() => {
		requestAnimationFrame(() => {
			const contentStartElement = document.getElementById("content-start");

			if (!contentStartElement) return true;

			contentStartElement.setAttribute("tabIndex", "-1");

			//scrolling to 0,0 to preserve existing BNF navigation behaviour
			window.scrollTo(0, 0);
			return false;
		});
	});
	// Default to scrolling to the content start element as the standard navigation behaviour

	return true;
};
