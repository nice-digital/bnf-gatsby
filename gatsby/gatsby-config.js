require("dotenv").config({
	path: `.env.${process.env.NODE_ENV}`,
});

const isBNF = process.env.GATSBY_SITE === "bnf";

if (!process.env.GATSBY_SEARCH_URL)
	throw new Error(
		"Env var GATSBY_SEARCH_URL isn't set. Did you forget to add it?"
	);

module.exports = {
	jsxRuntime: "automatic",
	siteMetadata: {
		isBNF,
		siteUrl: `https://${isBNF ? "bnf" : "bnfc"}.nice.org.uk`,
		siteTitleShort: isBNF ? "BNF" : "BNFC",
		siteTitleLong:
			"British National Formulary" + (isBNF ? "" : " for Children"),
	},
	plugins: [
		{
			resolve: `gatsby-source-bnf`,
			options: {
				feedURL: process.env.FEED_URL,
				userKey: process.env.FEED_USER_KEY,
				site: isBNF ? "bnf" : "bnfc",
			},
		},
		// Avoid errors like "ModuleNotFoundError: Module not found: Error: Can't resolve '@/components/Layout/Layout'" when using custom paths in tsconfig.json
		`gatsby-plugin-tsconfig-paths`,
		{
			resolve: `gatsby-plugin-sass`,
			options: {
				// Make CSS modules play nicely with TypeScript, see also global.d.ts for *.scss TS def
				cssLoaderOptions: {
					esModule: false,
					modules: {
						namedExport: false,
					},
				},
			},
		},
		"gatsby-plugin-catch-links",
		"gatsby-plugin-react-helmet",
		"gatsby-plugin-remove-generator",
		{
			resolve: `gatsby-plugin-sitemap`,
			options: {
				excludes: [`/search/`],
			},
		},
		{
			resolve: "gatsby-plugin-google-tagmanager",
			options: {
				id: "GTM-5H5L9GK",
				includeInDevelopment: true,
			},
		},
		{
			resolve: "gatsby-plugin-manifest",
			options: {
				icon: "src/images/icon.png",
			},
		},
	],
};
