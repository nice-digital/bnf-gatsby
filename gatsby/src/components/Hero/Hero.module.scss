@use '../../styles/colours';
@use '@nice-digital/nds-core/scss/layout';
@use '@nice-digital/nds-core/scss/media-queries';
@use '@nice-digital/nds-core/scss/spacing';
@use '@nice-digital/nds-core/scss/typography';
@use '@nice-digital/nds-core/scss/utils';

.hero {
  background: colours.$hero-background;
  left: 50%;
  margin-bottom: utils.rem(spacing.$large);
  margin-left: -50vw;
  margin-right: -50vw;
  overflow: hidden;
  padding-bottom: utils.rem(spacing.$large);
  position: relative;
  right: 50%;
  width: 100vw;

  .heroContainer {
    @include layout.container;
  }

  .title {
    @include typography.h1;
  }

  .intro {
    @include typography.lead;
    max-width: unset;
  }

  .lastUpdated {
    h2 {
      @include typography.h5;
      margin: 0;
    }

    time {
      @include typography.h3;
      display: block;
      font-weight: 700;
      margin: 0 0 utils.rem(spacing.$medium);
    }
  }
}

.breadcrumbs {
  flex: 0 1 100%;
}

@include media-queries.mq($from: md) {
  .heroContainer {
    align-items: center;
    column-gap: utils.rem(spacing.$medium);
    display: flex;
    flex-flow: row;
    flex-wrap: wrap;
  }

  .text {
    flex-grow: 1;
  }

  .lastUpdated {
    background-color: colours.$hero-last-updated-background;
    padding: utils.rem(spacing.$medium);
    padding-left: utils.rem(spacing.$medium);
    padding-right: utils.rem(spacing.$large);
  }

  .lastUpdated * {
    white-space: nowrap;
  }
}
