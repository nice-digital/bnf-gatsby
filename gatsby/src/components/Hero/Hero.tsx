import React from "react";

import styles from "./Hero.module.scss";

type HeroProps = {
	id: string;
	title: string;
	intro: string;
};

export const Hero: React.FC<HeroProps> = (props: HeroProps) => {
	const { id, title, intro } = props;
	return (
		<div className={styles.hero} id={id}>
			<div className={styles.hero_container}>
				<div className={styles.hero__body}>
					<div className={styles.hero__copy}>
						<h1 className={styles.hero__title}>{title}</h1>
						{intro && <p className={styles.hero__intro}>{intro}</p>}
					</div>
				</div>
			</div>
		</div>
	);
};
