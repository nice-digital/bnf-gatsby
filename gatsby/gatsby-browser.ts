import { RouteUpdateArgs } from "gatsby";

// Gatsby hook for when the route has changed on the client side
// See https://www.gatsbyjs.org/docs/browser-apis/#onRouteUpdate
export const onRouteUpdate = ({
	prevLocation,
	location,
}: RouteUpdateArgs): void => {
	if (prevLocation) {
		// Cast is needed because of https://github.com/gatsbyjs/gatsby/issues/29124
		const prevPath = (prevLocation as Location).pathname,
			path = location.pathname;

		// Push our own event to the dataLayer on page change rather than using the
		// 'gatsby-route-change' event built into gatsby-plugin-google-tagmanager.
		// Because gatsby-route-change is pushed on initial page load as well as route change,
		// AND because it doesn't use requestAnimationFrame to delay until the page title has been updated

		const sendPageView = () => {
			window.dataLayer.push({
				location: location.href,
				referrer: prevLocation.href,
			});
			// Don't consider hash changes to be a page view - pageviews only happy when the path changes
			if (prevPath != path) window.dataLayer.push({ event: "pageview" });
		};

		// Delay before push to the data layer, to make sure the page title has been updated
		// See https://github.com/gatsbyjs/gatsby/pull/10917/files#diff-bf0d94c8bf47d5c1687e342c2dba1e00R12-R13
		if ("requestAnimationFrame" in window) {
			window.requestAnimationFrame(() => {
				window.requestAnimationFrame(sendPageView);
			});
		} else {
			// simulate 2 rAF calls
			setTimeout(sendPageView, 32);
		}
	} else {
		window.dataLayer.push({
			location: location.href,
			referrer: document.referrer,
		});
	}
};
