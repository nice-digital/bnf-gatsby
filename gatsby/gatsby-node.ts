import { CreateSchemaCustomizationArgs, CreateWebpackConfigArgs } from "gatsby";

import { initialFieldExtension } from "./src/field-extensions/initial";

/**
 * Gatsby hook for customizing the schema.
 * See https://www.gatsbyjs.org/docs/schema-customization/
 */
export const createSchemaCustomization = ({
	actions: { createFieldExtension, createTypes },
}: CreateSchemaCustomizationArgs): void => {
	createFieldExtension(initialFieldExtension);

	// Add initials for types used in A-Z lists, as it's easier for grouping
	// We don't add the initial field in the source plugin because grouping is a presentation concern
	const typeDefs = `
	  type BnfDrug implements Node {
		"The lowercase first letter of the title, used for grouping"
		initial: String! @initial(field: "title")
	  }
	  type BnfTreatmentSummary implements Node {
		"The lowercase first letter of the title, used for grouping"
		initial: String! @initial(field: "title")
	  }
	  type BnfInteractant implements Node {
		"The lowercase first letter of the title, used for grouping"
		initial: String! @initial(field: "title")
	  }
	`;
	createTypes(typeDefs);
};

/**
 * Gatsby hook for overriding webpack config
 * See https://www.gatsbyjs.com/docs/how-to/custom-configuration/add-custom-webpack-config/
 */
export const onCreateWebpackConfig = ({
	stage,
	actions,
	getConfig,
	plugins,
}: CreateWebpackConfigArgs): void => {
	// Disable order warnings from mini CSS plugin
	// See https://github.com/gatsbyjs/gatsby/discussions/30169#discussioncomment-621285
	if (stage === "build-javascript" || stage === "develop") {
		const config = getConfig();

		const miniCss = config.plugins.find(
			(plugin: () => void) => plugin.constructor.name === "MiniCssExtractPlugin"
		);

		if (miniCss) {
			miniCss.options.ignoreOrder = true;
		}

		actions.replaceWebpackConfig(config);

		actions.setWebpackConfig({
			plugins: [plugins.provide({ process: "process/browser" })],
		});
	}
};
