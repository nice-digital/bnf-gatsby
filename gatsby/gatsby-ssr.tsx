import { PreRenderHTMLArgs } from "gatsby";
import React from "react";

export const onPreRenderHTML = ({
	getHeadComponents,
	replaceHeadComponents,
}: PreRenderHTMLArgs): void => {
	const components = getHeadComponents();

	components.push(
		<script
			src={process.env.GATSBY_COOKIE_BANNER_URL}
			type="text/javascript"
			key="cookie-banner"
			async
		></script>
	);

	replaceHeadComponents(components);
};
