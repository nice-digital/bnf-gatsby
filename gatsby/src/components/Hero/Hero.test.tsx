import { render, screen } from "@testing-library/react";
import * as Gatsby from "gatsby";

import { Hero, type LastUpdatedDataQueryResult } from "@/components/Hero/Hero";

const metaDataQueryResult: LastUpdatedDataQueryResult = {
	bnfMetadata: {
		lastUpdatedDate: "2022-04-06T00:38:57.911",
		lastUpdatedDateFormatted: "6 April 2022",
		runTag: "2022_04_06_0037_bnf",
	},
};

const useStaticQuery = jest.spyOn(Gatsby, "useStaticQuery");

beforeAll(() => {
	useStaticQuery.mockImplementation(() => metaDataQueryResult);
});

beforeEach(() => {
	jest.clearAllMocks();
});

describe("Hero", () => {
	it("should match snapshot for BNF", () => {
		render(<Hero isBNF={true} />);
		screen.getByRole("heading", {
			name: /british national formulary \(bnf\)/i,
		});
	});

	it("should match snapshot for BNFC", () => {
		render(<Hero isBNF={false} />);
		expect(
			screen.getByRole("heading", {
				name: /british national formulary for children \(bnfc\)/i,
			})
		).toMatchSnapshot();
	});

	it.only("should match snapshot for query", async () => {
		render(<Hero isBNF={false} />);
		const calls = useStaticQuery.mock.calls[0][0];
		expect(calls).toMatchInlineSnapshot(`
		"
			query LastUpdatedQuery {
				bnfMetadata {
					lastUpdatedDateFormatted: exportStarted(formatString: \\"D MMMM YYYY\\")
					lastUpdatedDate: exportStarted
					runTag
				}
			}
		"
	`);
		expect(useStaticQuery).toMatchSnapshot();
	});
});
