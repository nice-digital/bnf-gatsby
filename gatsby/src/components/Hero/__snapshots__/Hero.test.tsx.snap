// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hero should match snapshot for BNFC 1`] = `
<h1
  class="title"
>
  British National Formulary for Children (BNFC)
</h1>
`;

exports[`Hero should match snapshot for query 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "
	query LastUpdatedQuery {
		bnfMetadata {
			lastUpdatedDateFormatted: exportStarted(formatString: \\"D MMMM YYYY\\")
			lastUpdatedDate: exportStarted
			runTag
		}
	}
",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "bnfMetadata": Object {
          "lastUpdatedDate": "2022-04-06T00:38:57.911",
          "lastUpdatedDateFormatted": "6 April 2022",
          "runTag": "2022_04_06_0037_bnf",
        },
      },
    },
  ],
}
`;
